src/mqueryfront/src/QueryPage.js

```diff
@@ -95,40 +98,48 @@ class QueryPage extends Component {
                     "/matches/" +
                     this.state.qhash +
                     "?offset=" +
-                    this.state.matches.length +
+                    OFFSET +
                     "&limit=" +
                     LIMIT
             )
             .then((response) => {
-                let newShouldRequest = true;
-
-                if (
-                    ["done", "cancelled", "failed", "expired"].indexOf(
-                        response.data.job.status
-                    ) !== -1
-                ) {
-                    if (!response.data.matches.length) {
-                        newShouldRequest = false;
-                    }
-                }
-
+                let job = response.data.job;
                 this.setState({
-                    matches: [...this.state.matches, ...response.data.matches],
-                    job: response.data.job,
+                    job: job,
+                    matches: response.data.matches,
                 });
-
-                if (newShouldRequest) {
-                    let nextTimeout =
-                        response.data.matches.length >= LIMIT ? 50 : 1000;
-                    this.timeout = setTimeout(
-                        () => this.loadMatches(),
-                        nextTimeout
-                    );
+                let doneStatuses = ["done", "cancelled", "failed", "expired"];
+                let isDone = doneStatuses.indexOf(job.status) !== -1;
+                let processedAll = job.files_processed >= job.total_files;
+                if (isDone && processedAll) {
```

Comment:
Still `&&`? If nothing else changed, this will cause infinite loop of requests.

Reply:
I've also confirmed it locally. Test plan:

- start a big query
- stop it in a middle such that files_processed < total_files (remember that they're strings)
- observe that requests are repeated in the loop every 1s